### ISSUE TITLE
monitor worker count = 1 bug or confusion with RedisStorage

### ISSUE THREAD

**Issue by alexlatif** [geofmureithi/apalis#367]
I expect this 

```rust
    let monitor = Monitor::<TokioExecutor>::new()
        .register_with_count(1, {
```

to enable only one worker and would not expect the process_job to start processing the second job whilst the first is still processing. yet instead 1 worker does not correctly make it synchronous.

```bash
2024-07-13T05:01:34.764019Z  INFO apalis_test: Worker [test-worker-0] started
2024-07-13T05:01:34.868223Z DEBUG job: apalis::layers::tracing::on_request: job.start
2024-07-13T05:01:34.868295Z ERROR job: apalis_test: foo - Start processing job ID: job1
2024-07-13T05:01:34.868320Z DEBUG job: apalis::layers::tracing::on_request: job.start
2024-07-13T05:01:34.868359Z ERROR job: apalis_test: foo - Start processing job ID: job2
2024-07-13T05:01:36.870256Z ERROR job: apalis_test: Job ID job2 done
2024-07-13T05:01:36.870256Z ERROR job: apalis_test: Job ID job1 done
2024-07-13T05:01:36.870855Z DEBUG job: apalis::layers::tracing::on_response: job.done done_in=2002ms result=()
2024-07-13T05:01:36.870871Z DEBUG job: apalis::layers::tracing::on_response: job.done done_in=2002ms result=()
```

As you can see with one worker, but jobs appear to start. I dont think there is relevance in the redis configs. here is the way to reproduce.

```rust
use apalis::layers::tracing::TraceLayer;
use apalis::prelude::*;
use apalis::redis::RedisStorage;
use serde::{Deserialize, Serialize};
use std::time::Duration;
use tokio::time::sleep;
use tracing::{error, info};
use tracing_subscriber;

#[derive(Debug, Serialize, Deserialize)]
pub struct TestJob {
    job_id: String,
}

impl TestJob {
    pub fn new(job_id: String) -> Self {
        Self { job_id }
    }
}

impl Job for TestJob {
    const NAME: &'static str = "RedisJob";
}

pub async fn process_job(job: TestJob) -> Result<(), apalis::prelude::Error> {
    error!("foo - Start processing job ID: {}", job.job_id);

    let _ = sleep(Duration::from_secs(2)).await;

    error!("Job ID {} done", job.job_id);

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt().with_env_filter("trace").init();

    let conn = apalis::redis::connect("redis://127.0.0.1/").await?;
    let job_queue = RedisStorage::<TestJob>::new(conn);

    job_queue.clone().push(TestJob::new("job1".into())).await?;
    job_queue.clone().push(TestJob::new("job2".into())).await?;

    let worker = WorkerBuilder::new("test-worker")
        .layer(TraceLayer::new())
        .with_storage(job_queue.clone())
        .build_fn(process_job);

    let monitor = Monitor::<TokioExecutor>::new()
        .register_with_count(1, worker)
        .on_event(|e| {
            let worker_id = e.id();
            match e.inner() {
                Event::Start => {
                    info!("Worker [{worker_id}] started");
                }
                Event::Error(e) => {
                    error!("Worker [{worker_id}] encountered an error: {e}");
                }
                Event::Exit => {
                    info!("Worker [{worker_id}] exited");
                }
                _ => {}
            }
        })
        .shutdown_timeout(Duration::from_secs(5))
        .run_with_signal(async {
            tokio::signal::ctrl_c().await?;
            info!("Monitor starting shutdown");
            Ok(())
        });

    monitor.await?;

    Ok(())
}
```

```toml
[dependencies]
serde = { version = "1.0.204", features = ["derive"] }
apalis = { version = "0.5.3", features = ["apalis-redis", "redis"] }
tokio = { version = "1.38.0", features = ["full"] }
tracing = "0.1.40"
tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
redis = "0.25.4"
```

```bash
rustup 1.27.1 (54dd3d00f 2024-04-24)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.79.0 (129f3b996 2024-06-10)`
```

Am I not understanding .register_with_count(1, worker) or is this a bug with the worker count?

**Comment by alexlatif** [comment]
solved [via](https://github.com/geofmureithi/apalis/discussions/41) - very confusing why a worker needs a ConcurrencyLimitLayer. Could you please share the reason why a worker will use a thread pool and what the point of setting workers in the first place are? As in im confusing by having to tweak both to achieve one job at a time.

**Comment by geofmureithi** [comment]
There are several key points to consider:

   - ConcurrencyLayer: This pertains to the number of jobs or tasks that can be executed simultaneously.
   - Worker with Count: This relates to parallelism. Without using a thread pool, it would imply that a worker can only handle one job at a time, which is not optimal for our purposes.

Additionally, Apalis leverages Tower to delegate control to the library user. If a service is blocked (e.g., due to rate limits or concurrency constraints), workers are unable to poll for new tasks.

In short workers can consume as much as they can and you get to control that with tower layers.
